File: 2021/julia/src/Day6.jl
module Day6

include("Aoc.jl")
using .Aoc

numbers = parse.(Int64, split(Aoc.input_line(6), ","))

function evolve(days)
    counts = zeros(Int64, 9)

    foreach(n -> counts[n] += 1, numbers)

    for d in 2:days
        b = counts[1]

        foreach(i -> counts[i] = counts[i+1], 1:8)
        counts[7] += b
        counts[9] = b
    end

    sum(counts)
end

part1() = evolve(80)
part2() = evolve(256)

end

File: 2021/typescript/src/days/day6.ts
import {Day, sum} from "../aoc";

export class Day6 extends Day {
    day = (): number => 6;

    part1 = () => this.cycle(80);
    part2 = () => this.cycle(256);
    private fish: number[];

    cycle = (days: number) => {
        const f = [...this.fish];
        for(let d = 0; d < days; d++) {
            const b = f[0];
            for(let i = 0; i < 8; i++) {
                f[i] = f[i + 1];
            }
            f[8] = b;
            f[6] += b;
        }
        return f.reduce(sum);
    };

    setup = () => {
        this.fish = Array.from({length: 9}).map(() => 0);
        this.input[0].split(',').map(Number).forEach(f => this.fish[f]++);
    };
}